csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,457 @@
+    // all uses of the entry has been encounter in this traversal, we can
+    // safely propagate it across. remove val from frontier map
+    frontier.erase(val);
+    return true;
+  }
+
+  // updating uses in frontier. return false since we are not yet ready to
+  // propagate
+  frontier[val] = current_use;
+  return false;
+}
+
+// replayConcretePad tries to replay a concretized PadOp on pad_tv. This
+// function allows merging multiple consecutive PadOp by stacking their pad
+// widths together as `vec_pad_widths`. This function assumes it has already
+// resolved the output iter_type for each IterDomain and provided as
+// `ref_iter_type`. The function returns the output from the replay PadOp.
+//
+// NOTE: this assumes all vec_pad_widths are positive entries so we don't need
+// to consider accumulating them changing the output iter_type.
+TensorView* replayConcretePad(
+    TensorView* pad_tv,
+    Val* pad_value,
+    const std::vector<std::vector<Val*>>& vec_pad_widths,
+    std::vector<IterDomain*> ref_iter_type) {
+  NVF_ERROR(pad_tv->getDataType().has_value(), "pad source dtype is missing");
+  const std::vector<IterDomain*> inp_dom =
+      TensorDomain::noReductions(pad_tv->getLogicalDomain());
+  const auto rank = inp_dom.size();
+
+  NVF_ERROR(
+      rank == ref_iter_type.size(),
+      "ref_iter_type does not have compatible size regarding pad_tv");
+
+  std::vector<Val*> merged_pad_widths;
+  merged_pad_widths.reserve(rank * 2);
+
+  NVF_ERROR(!vec_pad_widths.empty(), "vec_pad_widths cannot be empty");
+  // stack vec_pad_widths as `merged_pad_widths`.
+  if (vec_pad_widths.size() == 1) {
+    merged_pad_widths = vec_pad_widths.at(0);
+  } else {
+    for (const auto i : c10::irange(2 * rank)) {
+      Val* merged_pad_width = nullptr;
+      for (const auto idx : c10::irange(vec_pad_widths.size())) {
+        // skipping zero pad;
+        Val* pad_width = vec_pad_widths[idx].at(i);
+        if (pad_width->isZeroInt()) {
+          continue;
+        }
+        merged_pad_width = merged_pad_width == nullptr
+            ? pad_width
+            : add(merged_pad_width, pad_width);
+      }
+      merged_pad_widths.push_back(
+          merged_pad_width == nullptr ? FusionGuard::getCurFusion()->zeroVal()
+                                      : merged_pad_width);
+    }
+  }
+
+  // construct TensorDomain for output TV.
+  std::vector<IterDomain*> merged_root_ids;
+  std::vector<IterDomain*> merged_logical_ids;
+  for (const auto i : c10::irange(rank)) {
+    Val* left_pad = merged_pad_widths.at(i * 2);
+    Val* right_pad = merged_pad_widths.at(i * 2 + 1);
+    IterDomain* inp_id = inp_dom.at(i);
+    if (left_pad->isZeroInt() && right_pad->isZeroInt()) {
+      merged_root_ids.push_back(inp_id->cloneWithoutRFactor());
+      merged_logical_ids.push_back(merged_root_ids.back());
+      continue;
+    }
+    // NOTE: nvfuser pad doesn't support negative padding, so we don't have to
+    // worry about it cancelling out.
```

Comment:
We do support pad by negative amounts, but it just won't happen in the `cat` scenario. See this test: https://github.com/NVIDIA/Fuser/blob/1bfb11a34d31200d7c2bc8a4e008a4279baf160b/tests/cpp/test_resize.cpp#L1666

Reply:
ha, I'll update and add a check for that.
Along with the comment here: https://github.com/NVIDIA/Fuser/blob/1416d8938fc3d3bd019c44cabb45d32d1ef944c3/csrc/ops/alias.cpp#L469-L470

Reply:
Yeah this doesn't look right when we try to merge multiple `PadOp` together. I'll address it in the follow up PR #2490  where we can actually put in proper tests.